File: gym/282/E.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
using ll = long long;

template<typename T> using V = vector<T>;
#define vi V<int>

struct DSU {
    vi e; void init(int n) { e = vi(n,-1); }
    int get(int x) { return e[x] < 0 ? x : e[x] = get(e[x]); }
    bool unite(int x, int y) {
        x = get(x), y = get(y); 
        if (x == y) return false;
        if (e[x]>e[y]) swap(x,y);
        e[x] += e[y]; e[y] = x;
        return true;
    }
};

struct edge {
    int u, v;
    ll w;
};

ll modpow(ll a, ll e, ll mod) { 
    ll r = 1;
    a %= mod;
    while (e) {
        if (e&1) r = r * a % mod;
        a = a * a % mod;
        e >>= 1;
    } 
    return r;
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr);
    int n, m; cin >> n >> m;
    vector<ll> A(n); for (int i = 0; i < n; ++i) cin >> A[i];
    vector<edge> edges;
    for (int i = 0; i < n; ++i) {
        for (int j = i+1; j < n; j++) { 
            ll w = (modpow(A[i], A[j], m) + modpow(A[j], A[i], m)) % m;
            edges.push_back({i,j,w});
        }
    }
    sort(edges.begin(), edges.end(), [&](auto& a, auto& b){ return a.w > b.w; });
    DSU d;
    d.init(n);
    ll mst_sum = 0;
    for (auto& e : edges) {
        if (d.unite(e.u, e.v)) mst_sum += e.w;
    }
    cout << mst_sum << '\n';
    return 0;
}

File: gym/282/E.py
import sys 
input = sys.stdin.readline

class DSU:
    def __init__(self, n):
        self.e = [-1]*n

    def find(self, x: int):
        if self.e[x] < 0: return x
        self.e[x] = self.find(self.e[x]) 
        return self.e[x]

    def union(self, x, y):
        x = self.find(x); y = self.find(y)
        if x == y: return False 
        if self.e[x] > self.e[y]: x, y = y, x 
        self.e[x] += self.e[y]
        self.e[y] = x
        return True 

def modpow(a, e, mod):
    r = 1
    a %= mod 
    while e:
        if e&1: r = r * a % mod 
        a = a * a % mod 
        e >>= 1
    return r 

n, m = map(int, input().split())
A = list(map(int,input().split()))
edges = []
for i in range(n):
    for j in range(i+1,n):
        w = (modpow(A[i], A[j], m) + modpow(A[j], A[i], m)) % m
        edges.append((i,j,w))
edges.sort(key=lambda x: x[2], reverse=True)
d = DSU(n)
mst_sum = 0
for u, v, w in edges:
    if d.union(u, v): mst_sum += w 
print(mst_sum)
